Mathlib/Data/List/GroupBy.lean

```diff
@@ -0,0 +1,156 @@
+  match l with
+  | nil => not_mem_nil _
+  | cons _ _ => nil_not_mem_groupBy_loop _ _ _ _ <| not_mem_nil _
+
+theorem ne_nil_of_mem_groupBy (r : α → α → Bool) {l : List α} (h : m ∈ l.groupBy r) : m ≠ [] := by
+  rintro rfl
+  exact nil_not_mem_groupBy r l h
+
+private theorem chain'_of_mem_groupBy_loop {r : α → α → Bool} {l : List α} {a : α} {g : List α}
+    (hga : ∀ b ∈ g.head?, r b a) (hg : g.Chain' (fun y x ↦ r x y)) {gs : List (List α)}
+    (hgs : ∀ m ∈ gs, m.Chain' (fun x y ↦ r x y)) (h : m ∈ groupBy.loop r l a g gs) :
+    m.Chain' (fun x y ↦ r x y) := by
+  revert a g gs
+  induction l with
+  | nil =>
+      intro a g hga hg gs hgs h
+      rw [groupBy.loop, reverse_cons, mem_append, mem_reverse, mem_singleton] at h
+      obtain hm | rfl := h
+      · exact hgs m hm
+      · apply List.chain'_reverse.1
+        rw [reverse_reverse]
+        exact chain'_cons'.2 ⟨hga, hg⟩
+  | cons b l IH =>
+      intro a g hga hg gs hgs h
+      simp [groupBy.loop] at h
+      split at h
+      · apply IH _ (chain'_cons'.2 ⟨hga, hg⟩) hgs h
+        intro b hb
+        rw [head?_cons, Option.mem_some_iff] at hb
+        rwa [← hb]
+      · apply IH _ chain'_nil _ h
+        · rintro _ ⟨⟩
+        · intro m
+          rw [mem_cons]
+          rintro (rfl | hm)
+          · apply List.chain'_reverse.1
+            rw [reverse_append, reverse_cons, reverse_nil, nil_append, reverse_reverse]
+            exact chain'_cons'.2 ⟨hga, hg⟩
+          · exact hgs _ hm
+
+theorem chain'_of_mem_groupBy {r : α → α → Bool} {l : List α} (h : m ∈ l.groupBy r) :
+    m.Chain' (fun x y ↦ r x y) := by
+  cases l with
+  | nil => cases h
+  | cons a l =>
+      apply chain'_of_mem_groupBy_loop _ _ _ h
+      · rintro _ ⟨⟩
+      · exact chain'_nil
+      · rintro _ ⟨⟩
+
+private theorem chain'_last_ne_head_groupBy_loop {r : α → α → Bool} (l : List α) {a : α}
+    {g : List α} {gs : List (List α)} (hgs' : [] ∉ gs)
+    (hgs : gs.Chain' (fun b a ↦ ∃ ha hb, r (a.getLast ha) (b.head hb) = false))
+    (hga : ∀ m ∈ gs.head?, ∃ ha hb, r (m.getLast ha) ((g.reverse ++ [a]).head hb) = false) :
+    (groupBy.loop r l a g gs).Chain' (fun a b ↦ ∃ ha hb, r (a.getLast ha) (b.head hb) = false) := by
+  revert a g gs
+  induction l with
+  | nil =>
+      intro a g gs _ hgs hga
+      rw [groupBy.loop, reverse_cons]
+      apply List.chain'_reverse.1
+      simpa using chain'_cons'.2 ⟨hga, hgs⟩
+  | cons b l IH =>
+      intro a g gs hgs' hgs hga
+      rw [groupBy.loop]
+      split
+      · apply IH hgs' hgs
+        intro m hm
+        obtain ⟨ha, _, H⟩ := hga m hm
+        refine ⟨ha, append_ne_nil_of_right_ne_nil _ (cons_ne_nil _ _), ?_⟩
+        rwa [reverse_cons, head_append_of_ne_nil]
+      · apply IH
+        · simpa using hgs'
+        · rw [reverse_cons]
+          apply chain'_cons'.2 ⟨hga, hgs⟩
+        · simpa
+
+theorem chain'_last_ne_head_groupBy (r : α → α → Bool) (l : List α) :
+    (l.groupBy r).Chain' (fun a b ↦ ∃ ha hb, r (a.getLast ha) (b.head hb) = false) := by
```

Comment:
You could probably give a version that doesn't require the existential, instead using `ne_nil_of_mem_groupBy`. However this might instead require using `.attach.Chain'`, so it wouldn't replace this one.

Reply:
Is it ok if I add this once we update to a new version of core Lean?

Reply:
(I asked this same question on Zulip and got yes for an answer, I've made PR #17044 which adds this relevant lemma to the bump branch)